### DISCUSSION TITLE
find and replace custom elements

### DISCUSSION THREAD

**Discussion by gaurishhs** [rehypejs/rehype#155] [Q&A]
hi! i'm making a utility to replace <outlook> tags with mso comments used by microsoft email engine (further a part of an email framework). it's work is to basically replace something like `<outlook>hi</outlook>` to `<!--[if mso]>hi<![endif]-->` i tried using `hast-util-find-and-replace` but that doesn't seem to work in my case. 

```
export const rehypeMso: Plugin<[Options?], Root, Root> = (options) => {
    if (!options) options = { tag: "outlook" };
    const processor = this;
    return (tree, vfile) => findAndReplace(tree, [
        [
            /outlook/gi, 
            ($0) => {
                return {
                    type: "comment",
                    value: "if !mso"
                }
            }
        ]
    ])
}   
```

i also tried to implement something like visiting every node and modifying them 

```ts
export const rehypeMso: Plugin<[Options?], Root, Root> = (options) => {
    if (!options) options = { tag: "outlook" };
    const processor = this;
    return (tree, vfile) => {
        const ctx: ComponentContext = { tree, vfile, processor };
        visit(tree, 'element', (node, index, parent) => {
            node.type = 'comment'
        })
    }
}   
```

but got typescript errors

Type '"comment"' is not assignable to type '"element"'.ts(2322) which is quite fair since i check for it to be an element first.

how should i be going about doing this?

**Comment by wooorm** [comment]
You seem to not be dealing with HTML, but perhaps some custom XML flavor.
These projects are for HTML.

`find-and-replace` replaces text. `<outlook>` is not text.

`visit` is perhaps what you want.
Perhaps our guides are also useful:
- https://unifiedjs.com/learn/recipe/tree-traversal/
- https://unifiedjs.com/learn/recipe/remove-node/
- https://unifiedjs.com/learn/recipe/tree-traversal-typescript/

Can you take some time investigating this info, change your example based on it, and post here if you still have questions?

**Reply by gaurishhs** [reply]
https://github.com/posthtml/posthtml-mso this does the exact same thing just that its for posthtml

**Reply by ChristianMurphy** [reply]
PostHTML can be used as a parser for `hast`/`rehype`. https://github.com/syntax-tree/hast/issues/3
There was a bridge to use PostHTML with retext (another unified project) https://github.com/posthtml/posthtml-retext but it is several major releases behind both retext and PostHTML, and is only a reference for how to bridge to `hast`.

**Reply by gaurishhs** [reply]
yeah i dont wanna use both at the same time.

**Reply by ChristianMurphy** [reply]
Okay, `rehype` is a general framework for working with HTML, it supports the `parse5` and DOM parsers, and can support others like PostHTML. "Both at the same time" doesn't really make sense as a characterization.

If you're looking for suggestions on how to leverage the `rehype` ecosystem, you've been given several.
If you don't want to use `rehype`, the `rehype` discussion forum may not be the best place to ask questions 😉😅 
If you have follow up questions on how to leverage `rehype` feel free to ask them.

**Reply by gaurishhs** [reply]
> "Both at the same time" doesn't really make sense as a characterization.

I apologise for the misunderstanding, i mean't that i don't want to use `posthtml` rather i would go with `parse5`

```ts
visit(tree, 'element', (node, index, parent) => {
            if (node.tagName === options?.tag) {
                // Replace it with mso comment
                if (!index) return;
                if (node.children.length !== 1) return;
                if (!parent || !parent?.children[index]) return;
                parent.children.splice(index, 1, {
                  type: "comment",
                  value: ` [if mso]> ${node.children[0].value} <![endif] `,
                })
                return [SKIP, index]                
            } 
        })
```
        
This is what i managed to do, This does work but i don't know whether i am doing it correctly or not, `node.children[0].value` should be used or not?

Input:

```ts
const file = await unified()
  .use(rehypeParse)
  .use(rehypeMso)
  .use(rehypeStringify).process(`<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Saturn</title>
  </head>
  <body>
    <h1>Saturn</h1>
    <outlook>
        hi 
        halo
        haahhaa
        a
        a
        a
        a
        a
        a
        
    </outlook>
  </body>
</html>`)
```

Output: 
```
<!doctype html><html lang="en"><head>
    <meta charset="utf-8">
    <title>Saturn</title>
  </head>
  <body>
    <h1>Saturn</h1>
    <!-- [if mso]> 
        hi 
        halo
        haahhaa
        a
        a
        a
        a
        a
        a
        
     <![endif] -->
  
</body></html>
```

**Reply by wooorm** [reply]
Not enough info.
See my earliest point: these projects are about HTML, old “IE-like comments” are not part of HTML.
Also, not enough info, you post certain input, and some code and some output which doesn’t work, and... What do you want? I don’t know what you want?

**Reply by gaurishhs** [reply]
Got that, Will try to be more informative next time. 

Thanks
